src/unix/kqueue.c

```diff
@@ -99,6 +99,44 @@ int uv__io_fork(uv_loop_t* loop) {
 int uv__io_check_fd(uv_loop_t* loop, int fd) {
   struct kevent ev;
   int rc;
+  struct stat sb;
+#ifdef F_GETPATH
+  char path[MAXPATHLEN];
+#endif
+
+  if(uv__fstat(fd, &sb)) {
+    rc = UV__ERR(errno);
+    return rc;
+  }
+
+  /* On FreeBSD, kqueue only supports EVFILT_READ notification for regular files
+   * and always reports ready events for writing, resulting in busy-looping.
+   *
+   * On Darwin, DragonFlyBSD, NetBSD and OpenBSD, kqueue reports ready events for
+   * regular files as readable and writable only once, acting like a EV_ONESHOT.
+   * 
+   * Neither of the above cases should be added to the kqueue.
+   */
+  if (S_ISREG(sb.st_mode) || S_ISDIR(sb.st_mode))
+    return UV_EINVAL;
+#ifdef __APPLE__
+  /* On Darwin (both macOS and iOS), in addition to regular files, FIFOs also don't
+   * work properly with kqueue: the disconnection from the last writer won't trigger
+   * an event for kqueue in spite of what the man pages say. Thus, we also disallow
+   * the case of S_IFIFO. */ 
+  if (S_ISFIFO(sb.st_mode)) {
+    rc = 0;
+    /* File descriptors of FIFO, pipe and kqueue share the same type of file, 
+     * therefore there is no way to tell them apart via stat.st_mode&S_IFMT.
+     * Fortunately, FIFO is the only one that has a persisted file on filesystem,
+     * from which we're able to make the distinction for FIFO. */
+#ifdef F_GETPATH
```

Comment:
When would this not be defined in Darwin?

Reply:
AFAIK, this macro normally should be present. I just went with this following the pattern of the existing code:
https://github.com/libuv/libuv/blob/5cc7175514571e41cc31219df1ae9dbd8e6b69f9/src/unix/kqueue.c#L490-L496
, only for overcaution.

Reply:
Just in case you aren't aware, minimum supported macOS version is 11.0, so "really old kernels" are of no concern.

Reply:
I'm okay with eliminating these `#if` directives if needed.

Reply:
I'd say let's do that.